kasenvr/project-athena
# Add Prometheus exporter

Exports domain statistics for the domain on port 9703 (officially reserved)

**What is this?**

This is a [Prometheus](https://prometheus.io/) exporter. It adds a webserver on port 9703 which delivers metrics about the domain (the same thing you can see in the domain's page) in a format Prometheus can consume.

**What is this for?**

It's for logging stats for graphing and troubleshooting. The end result will be a pretty graph like the one at the Hub. The initial use for this is troubleshooting audio issues -- seeing who loses packets, measuring bandwidth usage, etc.

**Testing**

1. Build a domain. There are changes both in the domain-server and assignment-clients.
2. Start the domain and go to Settings -> Monitoring to enable the exporter.
3. Verify nothing explodes when one goes to [http://localhost:9703](http://localhost:9703). The result should be a very simple explanatory page.
4. Go to [http://localhost:9703/metrics](http://localhost:9703/metrics). The result should be a list of metrics in a text format.
5. Optionally, verify that a Prometheus server can parse this. This will require installing Prometheus.
6. Optionally, make graphs with something like Grafana.
7. Optionally, verify that the stats are corrected tagged as "gauge" and "counter". Gauges are things that can go up and down, such as temperature or the current bandwidth usage. Counters are things that only increase, such as the number of requests that have been serviced.

This change should have no actual effect on a domain.
